### DISCUSSION TITLE
Which **tud_descriptor_** callbacks to implement?

### DISCUSSION THREAD

**Discussion by bkaz4283** [hathach/tinyusb#764] [Q&A] [answered]
Hello,

I am trying to implement MIDI over USB in my product using TinyUSB.  The "Getting Started" instructions say to "implement all **tud_descriptor_** callbacks..."  But how do I know which callbacks to implement exactly?

usbd.c contains the "process_get_descriptor()" function which anticipates six possibilities:
TUSB_DESC_DEVICE
TUSB_DESC_BOS
TUSB_DESC_CONFIGURATION
TUSB_DESC_STRING
TUSB_DESC_DEVICE_QUALIFIER
TUSB_DESC_OTHER_SPEED_CONFIG

However, in tusb_types.h I find the following additional descriptor types:
  TUSB_DESC_INTERFACE             = 0x04,
  TUSB_DESC_ENDPOINT              = 0x05,
  TUSB_DESC_INTERFACE_POWER       = 0x08,
  TUSB_DESC_OTG                   = 0x09,
  TUSB_DESC_DEBUG                 = 0x0A,
  TUSB_DESC_INTERFACE_ASSOCIATION = 0x0B,
  TUSB_DESC_DEVICE_CAPABILITY     = 0x10,
  TUSB_DESC_FUNCTIONAL            = 0x21,
  TUSB_DESC_CS_DEVICE             = 0x21,
  TUSB_DESC_CS_CONFIGURATION      = 0x22,
  TUSB_DESC_CS_STRING             = 0x23,
  TUSB_DESC_CS_INTERFACE          = 0x24,
  TUSB_DESC_CS_ENDPOINT           = 0x25,
  TUSB_DESC_SUPERSPEED_ENDPOINT_COMPANION     = 0x30,
  TUSB_DESC_SUPERSPEED_ISO_ENDPOINT_COMPANION = 0x31

A few questions:
1) How do I know which of these descriptors I need to support?  My device is only USB 1.1 compliant.
2) Do I need to add cases to the process_get_descriptor() function for the additional descriptors I need to support, basically copying the code for the other descriptors and modifying it as needed?
3) Is this what is meant by "implement all **tud_descriptor_** callbacks..."??

Please excuse my ignorance, I am really new to programming the USB stack and it all feels a bit overwhelming for now...

Thanks for any help you can offer!
-Brian

**Comment by hathach** [comment, marked as answer]
have you tried to run any of the stock examples, if not start & get familiar from there https://github.com/hathach/tinyusb/tree/master/examples/device
